My asynchronous

kazuma-naka opened the pull request.

kazuma-naka committed:
sqlite3 をインストール

kazuma-naka committed:
BooksDatabase クラスを作成。Booksテーブルを作成、削除、データの挿入、データの表示そしてデータベースをクローズのメソッドを作成

kazuma-naka committed:
BooksDatabase.js を 04.class から 03.asynchronous に移動

kazuma-naka committed:
BooksDatabase.js を 04.class ディレクトリで削除

kazuma-naka committed:
setTimeout を使って非同期処理をコールバックで行う

kazuma-naka committed:
Promise を使って非同期処理をするメソッドの実装。エラー無し

kazuma-naka committed:
async await の関数を追加。エラー無し

kazuma-naka committed:
エラーを含むデータ挿入とデータの表示のメソッドをコールバック、 Promise で作成。

kazuma-naka committed:
setTimeout を使用してエラーの関数の後にエラー無しの関数を実行する

kazuma-naka committed:
prettier を実行する

kazuma-naka committed:
使っていない関数に/* eslint-disable no-unused-vars */を追加して eslint を通す

kazuma-naka committed:
database.js を追加。コールバック形式、エラー無しの実装。

kazuma-naka committed:
SQL の構文にエラーを交えてコールバックで実装する

kazuma-naka committed:
コールバック形式をPromise でラップする。 prototype を使用する

kazuma-naka committed:
Promise でラップして then でつなげる。エラーあり

kazuma-naka committed:
async await に書き換えた関数を実装

kazuma-naka committed:
不必要な js ファイルを削除

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+#! /usr/bin/env node
```
`#!` の後ろのスペースはなくて良いと思います (あっても動きますが) 。https://en.wikipedia.org/wiki/Shebang_(Unix)#Syntax

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+#! /usr/bin/env node
+
+import sqlite3 from "sqlite3";
+import timers from "timers/promises";
```
import は、

1. Node.js の標準ライブラリ
2. NPM パッケージ
3. 自分で定義したファイル (今回はないですが)

の順に書くのがよいと思います。

cafedomancer commented on 03.asynchronous/database.js:
データベースのことだけを扱うプログラムというわけではないので、`database.js` というファイル名には違和感があります。

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+        id INTEGER PRIMARY KEY AUTOINCREMENT,
+        title TEXT NOT NULL UNIQUE
+        );
+    `;
```
この書き方だと以下のように余分な空白が多く含まれてしまいます。実行上は問題ないのですが好ましくはないので、1 行で書くか、もしくは [dedent](https://www.npmjs.com/package/dedent) などのライブラリを使って書いたほうがよいと思います。

```

        CREATE TABLE IF NOT EXISTS books (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        title TEXT NOT NULL UNIQUE
        );
    
```

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+import timers from "timers/promises";
+
+const database = new sqlite3.Database(":memory:");
+const createtableSQL = `
```
単語の区切り的に `createTableSQL` だと思います。

```suggestion
const createTableSQL = `
```

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+
+const database = new sqlite3.Database(":memory:");
+const createtableSQL = `
+        CREATE TABLE IF NOT EXISTS books (
```
今回はインメモリデータベースを使っており、この SQL の実行時点でテーブルが存在している可能性はないので、`IF NOT EXISTS` をつけても特に意味がないと思います。`DROP TABLE` の `IF EXISTS` についても同様です。

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+        );
+    `;
+
+const insertSQL = `INSERT INTO books (title) VALUES (?)`;
```
文字列中に変数を埋め込んでいないのにテンプレート文字列が使われています。

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+const selectSQL = "SELECT id, title FROM books";
+const dropTableSQL = `DROP TABLE IF EXISTS books`;
+const insertSQLError = `INSERT INTO boooks (title) VALUES (?)`;
+const selectSQLError = "SELECT id, tile FROM books";
```
第一義は SQL であることなので、変数名が `Error` で終わっているのは適当ではなさそうです。

cafedomancer replied:
7 行目から 18 行目までで一連の変数定義に見えるので、13 行目に空行があるのは違和感があります。

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+const dropTableSQL = `DROP TABLE IF EXISTS books`;
+const insertSQLError = `INSERT INTO boooks (title) VALUES (?)`;
+const selectSQLError = "SELECT id, tile FROM books";
+const titleNames = [
```
カラム名と合わせて `titles` で良いんじゃないでしょうか。

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+
+/* callback start */
+database.run(createtableSQL, (err) => {
+  if (err) return console.error(err);
```
エラーを起こさない想定の箇所のエラーハンドリングは不要です。また、if 文は常にブロックを伴って書いてください　(https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Statements/if...else#%E8%A7%A3%E8%AA%AC) 。また、`console.error()` は戻り値のない関数なので、その戻り値を return しても意味がないと思います。

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+  console.log("Books のテーブルを作成");
+  database.run(insertSQL, [titleNames[0]], function (err) {
+    if (err) return console.error(err);
+    console.log(`${titleNames[0]} が挿入されました。\nrowid: ${this.lastID}`);
```
改行を使って 1 つの文字列にするのではなく、2 つの `console.log` の実行に分けたほうが分かりやすいと思います。ところで、`rowid` と `id` は異なるものなので、`id` を `rowid` として出力するのは正しくないと思います。

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+  });
+} catch (err) {
+  console.error(err);
+}
```
コールバック関数で扱われているのは例外ではないので、これらの処理を try...catch で囲んでも意味がないと思います。

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+    if (err) return console.error(err);
+    console.log("Books のテーブルを作成");
+    database.run(insertSQLError, [titleNames[0]], function (err) {
+      if (err) return console.error(err);
```
この実装だと後続の処理が実行されなくなってしまうと思います。エラーの有無によらずに後続の処理 (Docs に記載のある 1-4 の処理) を実行するようにしてください。また、標準エラー出力にはエラーそのものではなくエラーのメッセージを出力してください。

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+      }
+    });
+  });
+};
```
`sqlite3.Database` に関数を定義するのではなく、トップレベルに関数を定義してください。そうする必要がある場合を除いて、ライブラリの既存の実装を差し替えるのは一般的にはあまり良くないアイディアだと思います。

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+          console.log(`${params} が挿入されました。\nrowid: ${this.lastID}`);
+        else if (sqlQuery === dropTableSQL)
+          console.log("テーブルを削除しました。");
+        else console.log("不正な SQL クエリです");
```
関数の再利用性の観点から、エラーがない場合にどのような処理を行うかは、promise を使う側が定められるようになっているべきです。

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+  } catch (error) {
+    console.error(error);
+  }
+})();
```
Top-level を使えば、await を使うために即時実行関数式を定義する必要はなくなると思います。

cafedomancer commented on 03.asynchronous/database.js:
```diff
@@ -0,0 +1,236 @@
+    await database.close();
+  } catch (error) {
+    console.error(error);
+  }
```
これだとあらゆる例外が捕捉されてしまいます。捕捉したい例外のみを捕捉するようにしてください。

kazuma-naka committed:
import する順番、ファイル名、create table のSQL のコマンドを１行にするなどのリファクタ

kazuma-naka committed:
非同期のコールバック、エラーありの処理のエラーをハンドリングして全て出力する。

kazuma-naka committed:
Promise でラップした run と each の関数を実装する

kazuma-naka committed:
プラクティスの1,2,3 どれを始めるかログに出力する関数を実装する

kazuma-naka committed:
runPromise をリファクタする

kazuma-naka committed:
Promise を await でエラー無し、エラーありで実装する

kazuma-naka committed:
コールバック形式のエラー無しをリファクタ

kazuma-naka committed:
変数名やテンプレート文字列など直す

kazuma-naka committed:
all を Promise でラップして保存されたすべてのid とtitle を表示する

kazuma-naka committed:
エラーの想定していない箇所のエラーハンドリングを削除」。if 文をブロックで囲う

kazuma-naka committed:
Console.log を二つに分けて見やすくする

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,242 @@
+  "The Great Gatsby",
+  "Pride and Prejudice",
+  "The Catcher in the Rye",
+];
```
main 関数相当のトップレベルの処理はそれ単体でまとめて書いてください。現状だと、変数の宣言、関数の宣言、処理の実行、の順に書かれていますが、変数の宣言と処理の実行は main 関数相当の処理なので、それらをまとめて書いてほしいです :pray: ただ、SQL を格納している変数は一種の定数のようにも見えるので、これらはファイルの上部に定義されていてもよいかもしれません。こちらをどうするかはお任せします。

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,242 @@
+                  console.log(`${row.id}: ${row.title}`);
+                } catch (err) {
+                  console.error(err.message);
+                }
```
try...catch を使わなくても同様の処理が行えるはずです。

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,242 @@
+      }
+      database.run(dropTableSQL, () => {
+        console.log("テーブルを削除しました。");
+      });
```
この処理が実行されるタイミングがエラーなしのプログラムと変わってしまっているような気がします。

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,242 @@
+      });
+    });
+  })
+  .catch((err) => console.error(err));
```
コールバックによるネストの問題を解決するために Promise を導入しているのに、その Promise がネストしまっているのはなにかおかしい気がします。

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,242 @@
+  await runPromise(database, insertErrorSQL, titles[0]);
+} catch (err) {
+  console.error(err.message);
+}
```
https://github.com/kazuma-naka/js-practices/pull/4#discussion_r1686191227 にコメントしている問題が解決していないです。

kazuma-naka committed:
callback のeach をall に変える

kazuma-naka committed:
callback のエラー有 SQLITE_ERROR のみをerr.code で取得する

kazuma-naka committed:
コールバックの dropTable のタイミングを合わせる

kazuma-naka committed:
Promise エラー無しをネストを解消してチェーンでつなげる。

kazuma-naka committed:
callback エラー無しの処理の順番を課題通りにする

kazuma-naka committed:
callback エラーありの順番をエラー無しと合わせる。特定のエラーをハンドルする関数を作成する

kazuma-naka committed:
Promise エラーありをチェーンでつなげエラーハンドリングもする

kazuma-naka committed:
await エラーありのエラーハンドリングの修正。エラー無しで５つのタイトルを挿入してすべて同じ処理にする。

kazuma-naka committed:
main 関数に相当する処理を上に移動

kazuma-naka committed:
callback エラー無しのタイトルを取得するように修正

kazuma-naka committed:
runPromise でデータが挿入される時などparams が空でない場合this.lastID をそれ以外は null を返すように変更する

kazuma-naka committed:
全ての出力が同じになるように修正する

kazuma-naka committed:
出力が完全に同じになるように修正する。

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,294 @@
+  console.log("\x1b[36m%s\x1b[0m", "-".repeat(30));
+  console.log("\x1b[36m%s\x1b[0m", `${startType} を開始`);
+  console.log("\x1b[36m%s\x1b[0m", "-".repeat(30));
+}
```
関数はその仕様よりも前に宣言してください。function 文では巻き上げが起こるので問題ありませんが、const で宣言した関数は巻き上げが起こらないのでエラーになります。

cafedomancer commented on 03.asynchronous/bookSql.js:
SQL だけが定義されているわけではないので、`bookSql.js` というファイル名も適切ではなさそうです。

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,294 @@
+/* callback start */
+showStart("callback");
+database.run(createTableSQL, () => {
+  console.log("Books のテーブルを作成");
```
他の箇所と文言が一致していなさそうです。テーブルのことを指しているのであれば、小文字の books を使うのが正しいと思います。56 行目についても同様です。

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,294 @@
+                console.log(`id: ${row.id} title: ${row.title}`);
+              }
+              database.run(dropTableSQL, () => {
+                console.log("テーブルを削除しました。");
```
他の箇所と文言が一致していなさそうです。70 行目についても同様です。

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,294 @@
+          database.run(insertSQL, [titles[4]], function () {
+            console.log(`id: ${this.lastID}`);
+            database.all(selectSQL, (_, rows) => {
+              for (let row of rows) {
```
再代入されない変数は const を使って宣言してください。

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,294 @@
+  return new Promise((resolve, reject) => {
+    if (typeof params === "function" || typeof params === "undefined") {
+      params = [];
+    }
```
Promise 版の関数はコールバック関数を引数に受けないはずなので、この分岐は必要ないと思います。

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,294 @@
+        resolve(params.length ? this.lastID : null);
+      } catch (err) {
+        reject(err);
+      }
```
前にもコメントしていますが、try...catch を使う必要はないです。例外を扱わないといけない箇所でのみ、try...catch を使ってください。https://github.com/kazuma-naka/js-practices/pull/4#discussion_r1689029965

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,294 @@
+        if (err) {
+          throw err;
+        }
+        resolve(params.length ? this.lastID : null);
```
元の run 関数では `this.lastID` 以外の値も結果として取得できていたはずです。API ドキュメントを確認し、元の関数の機能を維持するような実装に修正してください。

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,294 @@
+function closePromise(db) {
+  return new Promise((resolve) => {
+    db.close();
+    resolve();
```
この実装だと `db.close()` が完了する前に promise が resolve されてしまいます。また、エラーが起こるケースが考慮されていません。main 相当の処理では、エラーが起こらない想定の箇所でエラーハンドリングを行う必要はありませんが、promise 版の関数では、元の関数の機能を維持する必要があるのでエラーを考慮した実装にしてください。

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,294 @@
+      console.error(err.message);
+    }
+  }
+}
```
以前にもコメントしていますが、https://github.com/kazuma-naka/js-practices/pull/4#discussion_r1686191227 にコメントしている問題が解決していないです。

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,294 @@
+
+function handleErrorSQL(err) {
+  if (err instanceof Error) {
+    if (err.code === "SQLITE_ERROR") {
```
この if 文は 1 つにまとめられそうです。

cafedomancer commented on 03.asynchronous/bookSql.js:
```diff
@@ -0,0 +1,294 @@
+} catch (err) {
+  handleErrorSQL(err);
+}
+
```
他のプログラムでは処理間に空行がなかったと思います。

kazuma-naka committed:
6回目のレビュー後のリファクタ

\nbookSql.js から bookAsync.js にファイル名を変更

\nConsole.log の出力を同じにする

for(let row of rows) から for(const row of rows) に変更

kazuma-naka committed:
Promise のリファクタ

try catch を無くす

closePromise をエラーハンドリングする

resolve で changes も取得できるようにする

kazuma-naka committed:
Promise を修正してSQLITE_ERROR のエラーのみを捕捉できるようにする

kazuma-naka committed:
main 関数で使用する関数を上部に持ってくる

cafedomancer commented on 03.asynchronous/bookAsync.js:
```diff
@@ -0,0 +1,275 @@
+const selectSQL = "SELECT DISTINCT id, title FROM books ORDER BY id";
+const dropTableSQL = "DROP TABLE books";
+const insertErrorSQL = "INSERT INTO boooks (title) VALUES (?)";
+const selectErrorSQL = "SELECT id, tile FROM books";
```
これらを定数として扱うのであれば、変数名には大文字の snake case を使ってください。

cafedomancer commented on 03.asynchronous/bookAsync.js:
```diff
@@ -0,0 +1,275 @@
+                console.log(`id: ${row.id} title: ${row.title}`);
+              }
+              database.run(dropTableSQL, () => {
+                console.log("テーブルを削除しました。");
```
https://github.com/kazuma-naka/js-practices/pull/4#discussion_r1694411247 の対応が漏れています。

cafedomancer commented on 03.asynchronous/bookAsync.js:
```diff
@@ -0,0 +1,275 @@
+function runPromise(db, sqlQuery, params = []) {
+  return new Promise((resolve, reject) => {
+    db.run(sqlQuery, params, function (err) {
+      if (err && err instanceof Error && err.code === "SQLITE_ERROR") {
```
https://github.com/kazuma-naka/js-practices/pull/4#discussion_r1694413699 に関してですが、ここでエラーの種類を絞り込んでも意味がないです。try...catch では、例外の種類によらずにあらゆる例外が捕捉されてしまい、例外を意図せずに握りつぶしてしまう可能性があるので、捕捉したい例外のみ捕捉するようにということをコメントしています。例えば、以下のプログラムでは、本来はプログラムを異常終了させるべきですが、try...catch で例外が不用意に扱われているため、そうなりません。

```js
try {
  console.log(nonExistentVariable.someProperty);
} catch (e) {
  console.error(e);
}
```

cafedomancer commented on 03.asynchronous/bookAsync.js:
```diff
@@ -0,0 +1,275 @@
+  })
+  .then(() => {
+    return runPromise(database, insertErrorSQL, titles[1]);
+  })
```
アロー関数の中身が return しかないときは、ブロック文体ではなく簡潔文体を使ってください。

cafedomancer commented on 03.asynchronous/bookAsync.js:
```diff
@@ -0,0 +1,275 @@
+showStart("await");
+await runPromise(database, createTableSQL);
+console.log("books テーブルを作成しました。");
+const insertTitle0 = await runPromise(database, insertSQL, titles[0]);
```
結果として得られるのは title そのものではないので、`insertTitle0` という変数名は適切ではなさそうです。

kazuma-naka committed:
変数名や出力のリファクタ

SQL 文の変数名に大文字の snake case に変更する

コールバックエラー無しのテーブルを削除した際の出力の修正

return のみのアロー関数を簡潔文体に変更

await エラー無しで INSERT_SQL を await で保持する変数名の変更

kazuma-naka committed:
await エラーあり、try catch エラーハンドリングの修正

SQLITE_ERROR の場合は Console.error でログを出力。それ以外はエラーを throw して異常終了するようにする。
